### PR TITLE
refactor daemoncontroller

### LINKED ISSUE #28839: defect of the daemonset
I built a  cluster of kubernetes with 5 hosts, and created a daemonset, while nodeselector and nodeName were not used. After the daemonset is created, I detected the amount of DESIRED was 4.( To my understanding, it should be 5).

Trouble-shooting: After an intensive troubleshooting, I found out that the amount of DESIRED is influenced by a function called nodeShouldRunDaemonPod. 
In general, this function is supposed to have an impact on the amount of DESIRED when one of following two cases occurs: the first case is the node doesn't satisfies the daemon set's node selector, or the nodeName is not match. which means the host should not run pod of the daemonset, while the second case is others such as insufficient free resources or host port conflict. which means the host should run pod but not now. However, in reality, the second case does not influence the amount of DESIRED.

Suggestion: I think the function nodeShouldRunDaemonPod should be split into two functions which are shouldRunDaemonPod and canRunDaemonPod respectively. Only the former one has an impact on the amount of DESIRED.

### PR DESCRIPTION
**Pull request by yarntime**
fix #28839.

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37803)

**Comment by yarntime** [comment]
I forget to think about the AffinityAnnotation. 
If this is a valid pr, I'll refactor it later, thanks.

**Comment by dims** [comment]
@k8s-bot ok to test

File: pkg/controller/daemon/daemoncontroller.go [hunk 1]
```diff
@@ -650,6 +652,22 @@ func (dsc *DaemonSetsController) nodeShouldRunDaemonPod(node *v1.Node, ds *exten
 		return false
 	}
 
+	newPod := &v1.Pod{Spec: ds.Spec.Template.Spec, ObjectMeta: ds.Spec.Template.ObjectMeta}
```
Comment (kargakis) [inline pkg/controller/daemon/daemoncontroller.go]: Make this an utility that accepts a daemonSet and a node name and it returns a Pod?

**Comment by yarntime** [comment]
@mikedanese , cloud you please take a look? thanks.

**Comment by mikedanese** [comment]
This is no longer needed

**Pull request closed without merging**